pkg/scheduler/queue/tree_queue.go

```diff
@@ -0,0 +1,282 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package queue
+
+import (
+	"container/list"
+)
+
+type QueuePath []string //nolint:revive
+type QueueIndex int     //nolint:revive
+
+const localQueueIndex = -1
+
+// TreeQueue is a hierarchical queue implementation with an arbitrary amount of child queues.
+//
+// TreeQueue internally maintains round-robin fair queuing across all of its queue dimensions.
+// Each queuing dimension is modeled as a node in the tree, internally reachable through a QueuePath.
+//
+// The QueuePath is an ordered array of strings describing the path through the tree to the node,
+// which contains the FIFO local queue of all items enqueued for that queuing dimension.
+//
+// When dequeuing from a given node, the node will round-robin equally between dequeuing directly
+// from its own local queue and dequeuing recursively from its list of child TreeQueues.
+// No queue at a given level of the tree is dequeued from consecutively unless all others
+// at the same level of the tree are empty down to the leaf node.
+type TreeQueue struct {
+	// name of the tree node will be set to its segment of the queue path
+	name                   string
+	maxQueueLen            int
+	localQueue             *list.List
+	currentChildQueueIndex int
+	childQueueOrder        []string
+	childQueueMap          map[string]*TreeQueue
+}
+
+func NewTreeQueue(name string, maxQueueLen int) *TreeQueue {
+	return &TreeQueue{
+		name:                   name,
+		maxQueueLen:            maxQueueLen,
+		localQueue:             list.New(),
+		currentChildQueueIndex: localQueueIndex,
+		childQueueMap:          map[string]*TreeQueue{},
+		childQueueOrder:        nil,
+	}
+}
+
+func (q *TreeQueue) IsEmpty() bool {
+	// avoid recursion to make this a cheap operation
+	//
+	// Because we dereference empty child nodes during dequeuing,
+	// we assume that emptiness means there are no child nodes
+	// and nothing in this tree node's local queue.
+	//
+	// In reality a package member could attach empty child queues with getOrAddNode
+	// in order to get a functionally-empty tree that would report false for IsEmpty.
+	// We assume this does not occur or is not relevant during normal operation.
+	return q.localQueue.Len() == 0 && len(q.childQueueMap) == 0
+}
+
+func (q *TreeQueue) NodeCount() int {
+	count := 1 // count self
+	for _, childQueue := range q.childQueueMap {
+		count += childQueue.NodeCount()
+	}
+	return count
+}
+
+func (q *TreeQueue) ItemCount() int {
```

Comment:
[nit] Similar to above, perhaps a name like `SubtreeQueueSize()` would make it clear this is referring not just to the size of this `TreeQueue`'s queue, but also all children and their children etc. as well.